python/lib/core/dmod/core/data_domain_detectors.py

```diff
@@ -0,0 +1,468 @@
+    def __init__(self, data_collection: DataCollection, collection_name: Optional[str] = None, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self._data_collection: DataCollection = data_collection
+        """ Collection of data items, analogous to a :class:`Dataset`, if not a dataset outright. """
+        self._collection_name: Optional[str] = collection_name
+        """
+        Optional name for collection, which is important when domains involve a ``data_id`` restriction.
+        
+        Note that **IFF** the data collection is a :class:`Dataset` object **AND** the passed ``collection_name`` param
+        is ``None``, then this attribute will be set to the ``name`` attribute of the dataset collection itself.
+        """
+        if collection_name is None and isinstance(data_collection, Dataset):
+            self._collection_name = data_collection.name
+
+        if isinstance(self._data_collection, Path) and not self._data_collection.is_dir():
+            raise ValueError(f"{self.__class__.__name__} initialized with a path require this path to be a directory.")
+        if isinstance(self._data_collection, Dataset) and self._data_collection.manager is None:
+            raise ValueError(f"Dataset used to initialize {self.__class__.__name__} must have manager set.")
+
+    def detect(self, **kwargs) -> DataDomain:
+        """
+        Detect and return the data domain.
+
+        Parameters
+        ----------
+        kwargs
+            Optional kwargs applicable to the subtype, which may enhance or add to the domain detection and generation
+            capabilities, but which should not be required to produce a valid domain.
+
+        Notes
+        -----
+        Detection is performed by merging individual item domains detected using :class:`UniversalItemDomainDetector`
+        instances.  This type does not influence the details of how individual domains detections are performed by
+        :class:`UniversalItemDomainDetector` objects.  The subsequent merging is performed by reducing the
+        individual item domains using :method:`DataDomain.merge_domains`. The order of the items processed when reducing
+        is based on the order of results of a call to :method:`get_item_names`.
+
+        Returns
+        -------
+        DataDomain
+            The detected domain.
+
+        Raises
+        ------
+        DmodRuntimeError
+            If it was not possible to properly detect the domain.
+        """
+        domain = reduce(DataDomain.merge_domains,
+                        {UniversalItemDomainDetector(item=self.get_item(i), item_name=i).detect() for i in
+                         self.get_item_names()})
+        # If this domain has a format with a self-reference to dataset id, and we have a name, then set that restriction
+        if StandardDatasetIndex.DATA_ID in domain.data_format.indices_to_fields().keys() and self._collection_name:
```

Comment:
Because a `DATA_ID` constraint never gets added to any item-specific domain.  A `data_id` is an intrinsic property of a _Dataset_, so the analog here is that it's a part of the collection.